cmd/bad-key-revoker/main.go

```diff
@@ -0,0 +1,360 @@
+	_, err := bkr.dbMap.Exec("UPDATE blockedKeys SET extantCertificatesChecked = true WHERE keyHash = ?", row.KeyHash)
+	return err
+}
+
+func (bkr *badKeyRevoker) resolveContacts(ids []int64) (map[int64]string, error) {
+	idToEmail := map[int64]string{}
+	for _, id := range ids {
+		var emails struct {
+			Contact []string
+		}
+		err := bkr.dbMap.SelectOne(&emails, "SELECT contact FROM registrations WHERE id = ?", id)
+		if err != nil {
+			return nil, err
+		}
+		if len(emails.Contact) == 0 {
+			continue
+		}
+		idToEmail[id] = strings.TrimPrefix(emails.Contact[0], "mailto:")
+	}
+	return idToEmail, nil
+}
+
+var emailTemplate = template.Must(template.New("email-template").Parse(`Hello,
+
+The public key associated with certificates which you have issued has been marked as compromised. As such we are required to revoke any certificates which contain this public key.
+
+The following currently unexpired certificates that you've issued contain this public key and have been revoked:
+{{range . -}}
+{{.}}
+{{end}}`))
+
+var emailSubject = "Certificates you've issued have been revoked due to key compromise"
+
+var maxSerials = 100
+
+func (bkr *badKeyRevoker) sendMessages(mapping map[string][]string) error {
+	err := bkr.mailer.Connect()
+	if err != nil {
+		return err
+	}
+	defer func() {
+		_ = bkr.mailer.Close()
+	}()
+	for email, serials := range mapping {
+		if len(serials) > maxSerials {
+			more := len(serials) - maxSerials
+			serials = serials[:maxSerials]
+			serials = append(serials, fmt.Sprintf("and %d more certificates.", more))
+		}
+		message := bytes.NewBuffer(nil)
+		err := emailTemplate.Execute(message, serials)
+		if err != nil {
+			return err
+		}
+		err = bkr.mailer.SendMail([]string{email}, emailSubject, message.String())
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+var keyCompromiseCode = int64(revocation.KeyCompromise)
+
+func (bkr *badKeyRevoker) invoke() (bool, error) {
+	// find rows to work on
+	uncheckedRows, err := bkr.selectUncheckedRows()
+	if err != nil {
+		return false, err
+	}
+	if len(uncheckedRows) == 0 {
+		return true, nil
+	}
+	for _, row := range uncheckedRows {
+		// find everything in keyHashToSerial with certificateStatus.status != revoked
+		unrevokedCerts, err := bkr.findUnrevoked(row)
+		if err != nil {
+			return false, err
+		}
+		if len(unrevokedCerts) == 0 {
+			// mark row as checked
+			err = bkr.markRowChecked(row)
+			if err != nil {
+				return false, err
+			}
+			continue
+		}
+		// revoke all rows, record who issued the cert
+		ownedBy := map[int64][]string{}
+		for _, cert := range unrevokedCerts {
+			_, err = bkr.raClient.AdministrativelyRevokeCertificate(context.Background(), &rapb.AdministrativelyRevokeCertificateRequest{
+				Cert:      cert.DER,
+				Code:      &keyCompromiseCode,
+				AdminName: &revokerName,
+			})
+			if err != nil {
+				return false, err
+			}
+			ownedBy[cert.RegistrationID] = append(ownedBy[cert.RegistrationID], cert.Serial)
```

Comment:
Having one loop that both (a) revokes the certs and (b) builds the map from account ID to serials makes it hard to reason about and hard to test (b) separately.

Since building that map is logically part of the next section, where we decide who to send email to, I would recommend moving it to the next stanza. You can even combine building the map form with building the list:

```
var ids []int64
var ownedBy map[string][]string{}
for cert := range unrevokedCerts {
 regID := cert.RegistrationID
 if ownedBy[regID] == nil {
  ids = append(ids, regID)
 }
 ownedBy[regID] = append(ownedBy[regID], cert.Serial
}
```

Also, broadly speaking, I think the "collect IDs and email addresses and then send email" section that comes below here could probably be made into its own function.